saasherder/changelog.py

```diff
@@ -0,0 +1,158 @@
+import yaml
+from itertools import chain
+import os
+import subprocess
+
+
+class Changelog:
+
+    workspace = "_workspace/"
+
+    @staticmethod
+    def context_dir(context):
+        return context + "-services/"
+
+    def path(self, context, service_name):
+        """Get YAML config path for a service"""
+
+        return self.context_dir(context) + service_name + ".yaml"
+
+    def definitions(self, context, service_name):
+        """Get list of services defined in service file"""
+
+        with open(self.path(context, service_name)) as f:
+            parsed = yaml.load(f)
+            return parsed['services']
+
+    def service_names(self, context):
+        """Get names of all services in a context"""
+
+        dir = self.context_dir(context)
+        files = os.listdir(dir)
+
+        return [os.path.splitext(f)[0] for f in files]
+
+    def services(self, context):
+        """Get all services in the context"""
+
+        names = self.service_names(context)
+        dsls = [self.definitions(context, name) for name in names]
+        return list(chain(*dsls))
+
+    @staticmethod
+    def diff(s1, s2):
+        """Return the difference between 2 service lists
+
+        This is not a commutative relationship. We are looking for services in
+        first list which differ from that in second list. a ⊕ b ≠ b ⊕ a.
```

Comment:
these special characters result in:

" File "../saasherder/saasherder/changelog.py", line 47
SyntaxError: Non-ASCII character '\xe2' in file ../saasherder/saasherder/changelog.py on line 48, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details"

Reply:
this was because I was running with `python` not `python3`

Reply:
Ah! Py 3 didn't give me a warning. I'll add the encoding header.